### PR TITLE
Unit test

### PR DESCRIPTION
**Pull request by MasonHN**

### CODE REVIEW

File: src/test/java/com/twitter/ResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @Test
+    public void testTimeline() {
+        Response responseTimeline = resourceMock.showTimeline();
+        assertEquals(Response.Status.OK.getStatusCode(), responseTimeline.getStatus());
+    }
+
+    @Test
+    public void testPost() {
+        Response responseTweet = resourceMock.createTweet(testTweet);
+        assertEquals(Response.Status.OK.getStatusCode(), responseTweet.getStatus());
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: check more than the status code

File: src/test/java/com/twitter/ResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.mockito.MockitoAnnotations;
+import twitter4j.TwitterException;
+import static org.mockito.Mockito.*;
+import javax.ws.rs.core.Response;
+import static org.junit.Assert.*;
+
+public class ResourceTest {
+
+    Tweet testTweet = new Tweet("tweet");
+    Tweet badTweet = new Tweet("tweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweettweet");
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: there is a better way than hard coding a string this long

File: src/test/java/com/twitter/ServicesTest.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import org.junit.Test;
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+import twitter4j.*;
+import java.util.List;
+import static org.mockito.Mockito.mock;
+public class ServicesTest {
+
+    @Test
+    public void servicesTest() throws TwitterException {
```
Comment (kev-chan) [inline src/test/java/com/twitter/ServicesTest.java]: this test needs work

File: src/main/java/com/twitter/Services/TwitterServices.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package com.twitter.Services;
```
Comment (kev-chan) [inline src/main/java/com/twitter/Services/TwitterServices.java]: all the refactoring was supposed to be done later as the focus of this lab was on unit testing the existing code but i'll let that go.

File: src/test/java/com/twitter/ResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
+package com.twitter;
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: use better variable names

File: src/test/java/com/twitter/ResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import twitter4j.TwitterException;
+import static org.mockito.Mockito.*;
+import javax.ws.rs.core.Response;
+import java.sql.Time;
+
+import static org.junit.Assert.*;
+
+public class ResourceTest {
+
+    Tweet testTweet = new Tweet("tweet");
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: no need to put these as global variables.  Especially the long tweet as it is only used once.

File: src/test/java/com/twitter/ResourceTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    String longTweet = new StringUtils().repeat("a", 281);
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: any reason this is hardcoded value for length?

File: src/test/java/com/twitter/ResourceTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    private TwitterClientResource resourceMock;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void testTimeline() throws TwitterException {
+        Response responseTimeline = resourceMock.showTimeline();
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: what does this actually return?

File: src/test/java/com/twitter/ResourceTest.java (hunk 5 above)
Comment (kev-chan) [reply src/test/java/com/twitter/ResourceTest.java]: you should be using mocking where applicable

File: src/test/java/com/twitter/ResourceTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        Response responseTweet = resourceMock.createTweet(badTweet);
+        Response expectedResponse = Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Please make sure your tweet is " + resourceMock.maxTweetSize +  " characters or less.").build();
+        assertEquals(expectedResponse.getStatus(), responseTweet.getStatus());
+        assertEquals(expectedResponse.getEntity(), responseTweet.getEntity());
+    }
+
+    @Test
+    public void testTimelineException() throws TwitterException {
+        when(resourceMock.showTimeline()).thenThrow(TwitterException.class);
+        Response expectedResponse = Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Please try your request again later").build();
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: why making an expectedResponse just to pull the values out later?

File: src/test/java/com/twitter/ServicesTest.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    @InjectMocks
+    private TwitterServices mockServices;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void timelineTest() throws TwitterException {
```
Comment (kev-chan) [inline src/test/java/com/twitter/ServicesTest.java]: what about negative test cases?

File: src/test/java/com/twitter/ResourceTest.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void testTimeline() throws TwitterException {
+        List<Status> mockedTimeline = mock(ResponseList.class);
+        when(serviceMock.getTimeline()).thenReturn(mockedTimeline);
+        Response showTimelineResponse = testResourceClass.showTimeline();
+        assertEquals(Response.status(Response.Status.OK).entity(new Timeline(mockedTimeline)).build().getStatus(), showTimelineResponse.getStatus());
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: again why build a Response object here just to pull out the status?

File: src/test/java/com/twitter/ResourceTest.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        assertEquals(Response.status(Response.Status.OK).entity(new Timeline(mockedTimeline)).build().getEntity().getClass(), showTimelineResponse.getEntity().getClass());
```
Comment (kev-chan) [inline src/test/java/com/twitter/ResourceTest.java]: same as above.  Also don't test the class but the actual object

File: src/test/java/com/twitter/ServicesTest.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
+package com.twitter;
```
Comment (kev-chan) [inline src/test/java/com/twitter/ServicesTest.java]: from a class name perspective, what if you had multiple service classes?  Would you have all the tests go here?  Same applies to ResourceTest

Comment (MasonHN) [reply src/test/java/com/twitter/ServicesTest.java]: I would have separate test classes to match each service class. I'll rename the test classes to be more explicit.

**Pull request merged**